Accessed resources

cryogenian opened the pull request.

cryogenian commented:
[ch4612]

rintcius commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -114,19 +132,22 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
+    ))).map(_ match {
+      case Left(e) => None
+      case Right(a) => Some(a)
+    })
```
⛳️ should be able to use `toOption` https://www.scala-lang.org/api/current/scala/util/Either.html#toOption:Option[B]

rintcius commented on build.sbt:
```diff
@@ -42,6 +43,10 @@ lazy val core = project
+
+      // mongo's `getDatabase` and `getCollection` have `Nullable` annotations and they raise
+      // warnings w/o this dependency
+      "com.google.code.findbugs"   %  "jsr305"                      % jsrVersion % Optional
```
I think this should be `provided` instead of `optional`? (i.e. available at compile time but not runtime)

djspiewak replied:
I would just remove the qualifier. There's no reason to use `"optional"` or `"provided"` unless writing a library, which this is not.

rintcius replied:
@djspiewak Oh you mean because it will be available at runtime anyway because it's using the datasource plugin mechanism (which doesn't recognize them)?

djspiewak replied:
Actually I just mean there's no reason *not* to make it a mandatory dependency. :-)

djspiewak commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -81,10 +84,17 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
-    observableAsStream(client.listDatabaseNames, DefaultQueueSize).map(Database(_))
+    observableAsStream(client.listDatabaseNames, DefaultQueueSize)
+      .map(Database(_))
+      .handleErrorWith { _ match {
```
FYI, the `_ match {` is unnecessary. Match bodies are actually partial functions, which are themselves function values.

djspiewak commented on datasource/src/main/scala/quasar/physical/mongo/Mongo.scala:
```diff
@@ -114,19 +132,22 @@ class Mongo[F[_]: MonadResourceErr : ConcurrentEffect] private[Mongo](client: Mo
 
     getStats map { doc =>
-      val optSize: Option[Long] = doc.get("avgObjSize").flatMap(_ match {
+      val optSize: Option[Long] = doc flatMap (_.get("avgObjSize") flatMap (_ match {
```
Stylistically, I would prefer:

```scala
doc.flatMap(_.get("avgObjSize")) flatMap {
```

Remember that monads are associative, so there's no need to nest the binds here.

djspiewak commented:
Left a few comments. Also added the ☠️ label since this is a really important thing to land, so feel free to ignore my comments until a follow-up PR if necessary. I'd really like to land this today or early tomorrow.

rintcius commented:
@cryogenian FYI nothing blocking from me either, given the ☠️ feel free to merge!

cryogenian commented:
@rintcius @djspiewak 

+ Modified tests to work with `--auth`ed mongo
+ Made `accessedResource` a method of `MongoConfig` 
+ Added handlers for constrained roles for 3.4 and >4.0 mongo versions for `databases` and `collections`

rintcius commented on datasource/src/test/scala/quasar/physical/mongo/MongoSpec.scala:
```diff
@@ -59,6 +56,16 @@ class MongoSpec extends EffectfulQSpec[IO] {
+    mongo.databases.compile.toList.map { _ === List(Database("B")) }})
+
+  "getting databases for constrained role raises an error if incorrect auth db specified" >> {
+    mkInvalidAMongo.unsafeRunSync() must throwA[Throwable]
```
this doesn't test what the description says (it mentions "getting databases")

rintcius commented on datasource/src/test/scala/quasar/physical/mongo/MongoSpec.scala:
```diff
@@ -162,6 +184,18 @@ object MongoSpec {
+          "createUser" -> "aUser",
+          "pwd" -> "aPassword",
+          "roles" -> List(Document("role" -> "read", "db" -> "A"))))).attempt
+        // bUser:bPassword --> can do anything with "B" roles, but can't do anything with data evel listCollections
```
evel -> level

rintcius commented on datasource/src/test/scala/quasar/physical/mongo/MongoSpec.scala:
```diff
@@ -128,11 +138,23 @@ object MongoSpec {
+  val connectionString: String = "mongodb://root:secret@127.0.0.1:27018"
+  val aConnectionString: String = "mongodb://aUser:aPassword@127.0.0.1:27018/A.a"
+  val invalidAConnectionString: String = "mongodb://aUser:aPassword@127.0.0.1:27018"
+  val bConnectionString: String = "mongodb://bUser:bPassword@127.0.0.1:27018/B.b"
```
also worth to test with a db only connection string (like `mongodb://bUser:bPassword@127.0.0.1:27018/B`) I think
